sklearn/feature_selection/tests/test_mutual_info.py

```diff
@@ -0,0 +1,173 @@
+from __future__ import division
+
+import numpy as np
+from numpy.testing import run_module_suite
+from scipy.sparse import csr_matrix
+
+from sklearn.utils.testing import (
+    assert_array_equal, assert_almost_equal, assert_true)
+from sklearn.feature_selection import MinRedundancyMaxRelevance
+from sklearn.feature_selection.mutual_info import _compute_mi
+
+
+class TestMIComputation(object):
+    def test_dd(self):
+        # In discrete case computations are straightforward and can be done
+        # by hand on given vectors.
+        x = np.array([0, 1, 1, 0, 0])
+        y = np.array([1, 0, 0, 0, 1])
+
+        H_x = H_y = -(3/5) * np.log(3/5) - (2/5) * np.log(2/5)
+        H_xy = -1/5 * np.log(1/5) - 2/5 * np.log(2/5) - 2/5 * np.log(2/5)
+        I_xy = H_x + H_y - H_xy
+
+        assert_almost_equal(_compute_mi(x, y, True, True), I_xy)
+
+    def test_cc(self):
+        # For two continuous variables a good approach is to test on bivariate
+        # normal distribution, where mutual information is known.
+
+        # Mean of the distribution, irrelevant for mutual information.
+        mean = np.zeros(2)
+
+        # Setup covariance matrix with correlation coeff. equal 0.5.
+        sigma_1 = 1
+        sigma_2 = 10
+        corr = 0.5
+        cov = np.array([
+            [sigma_1**2, corr * sigma_1 * sigma_2],
+            [corr * sigma_1 * sigma_2, sigma_2**2]
+        ])
+
+        # True theoretical mutual information.
+        I_theory = (np.log(sigma_1) + np.log(sigma_2) -
+                    0.5 * np.log(np.linalg.det(cov)))
+
+        np.random.seed(0)
+        Z = np.random.multivariate_normal(mean, cov, size=1000)
+
+        x, y = Z[:, 0], Z[:, 1]
+        I_computed = _compute_mi(x, y, False, False)
+
+        # Theory and computed values won't be very close, assert that relative
+        # error is less than 10%.
+        assert_true(np.abs(I_computed - I_theory) < 0.1 * I_theory)
+
+    def test_cd(self):
+        # To test define a joint distribution as follows:
+        # p(x, y) = p(x) p(y | x)
+        # X ~ Bernoulli(0.5)
+        # (Y | x = 0) ~ Uniform(-1, 1)
+        # (Y | x = 1) ~ Uniform(0, 2)
+
+        # Use the following formula for mutual information:
+        # I(X; Y) = H(Y) - H(Y | X)
+        # Two entropies can be computed by hand:
+        # H(Y) = 3/2 * ln(2)
+        # H(Y | X) = ln(2)
+        # So finally we have:
+        # I(X; Y) = 1/2 * ln(2)
+
+        # Now we need to implement sampling from out distribution, which is
+        # done easily using conditional distribution logic.
+
+        n_samples = 1000
+        np.random.seed(0)
+        Z = np.empty((n_samples, 2))
+        Z[:, 0] = np.random.uniform(size=n_samples) > 0.5
+        mask = Z[:, 0] == 0
+        Z[mask, 1] = np.random.uniform(-1, 1, size=np.sum(mask))
+        mask = ~mask
+        Z[mask, 1] = np.random.uniform(0, 2, size=np.sum(mask))
+
+        x = Z[:, 0].astype(int)
+        y = Z[:, 1]
```

Comment:
I don't think it is necessary to create Z. let's just do

    y = np.empty(n_samples)
    mask = np.random.uniform() > p
    y[mask] = uniform(1, -1, size(mask))
    y[~mask] = uniform(0, 2, size(~mask))
    x = x.astype(int)

Makes the code slightly easier to follow.